src/coreclr/src/tools/Common/TypeSystem/Common/Utilities/GCPointerMap.cs

```diff
@@ -0,0 +1,272 @@
+
+            if (numPointerSizedCells > 0)
+            {
+                // Given the number of cells, how many Int32's do we need to represent them?
+                // (It's one bit per cell, but this time we need to round up.)
+                _gcFlags = new int[((numPointerSizedCells - 1) >> 5) + 1];
+            }
+            else
+            {
+                // Not big enough to fit even a single pointer.
+                _gcFlags = Array.Empty<int>();
+            }
+
+            _pointerSize = pointerSize;
+
+            _delta = 0;
+            _limit = numBytes;
+        }
+
+        public void MarkGCPointer(int offset)
+        {
+            Debug.Assert(offset >= 0);
+
+            int absoluteOffset = _delta + offset;
+
+            Debug.Assert(absoluteOffset % _pointerSize == 0);
+            Debug.Assert(absoluteOffset <= (_limit - _pointerSize));
+
+            int cellIndex = absoluteOffset / _pointerSize;
+
+            _gcFlags[cellIndex >> 5] |= 1 << (cellIndex & 0x1F);
+        }
+
+        public GCPointerMapBuilder GetInnerBuilder(int offset, int size)
+        {
+            Debug.Assert(offset >= 0);
+
+            int absoluteOffset = _delta + offset;
+
+            Debug.Assert(absoluteOffset + size <= _limit);
+
+            return new GCPointerMapBuilder
+            {
+                _gcFlags = this._gcFlags,
+                _pointerSize = this._pointerSize,
+                _delta = absoluteOffset,
+                _limit = absoluteOffset + size
+            };
+        }
+
+        public GCPointerMap ToGCMap()
+        {
+            Debug.Assert(_delta == 0);
+            return new GCPointerMap(_gcFlags, (_limit + _pointerSize - 1) / _pointerSize);
+        }
+
+        public BitEnumerator GetEnumerator()
+        {
+            int numCells = (_limit - _delta) / _pointerSize;
+            int startCell = _delta / _pointerSize;
+            return new BitEnumerator(_gcFlags, startCell, numCells);
+        }
+
+        public override string ToString()
+        {
+            var sb = new StringBuilder();
+            foreach (var bit in this)
+                sb.Append(bit ? '1' : '0');
+            return sb.ToString();
+        }
+    }
+
+    public struct BitEnumerator
+    {
+        private int[] _buffer;
+        private int _limitBit;
+        private int _currentBit;
+
+        public BitEnumerator(int[] buffer, int startBit, int numBits)
+        {
+            Debug.Assert(startBit >= 0 && numBits >= 0);
+            Debug.Assert(startBit + numBits <= buffer.Length << 5);
+
+            _buffer = buffer;
+            _currentBit = startBit - 1;
+            _limitBit = startBit + numBits;
+        }
+
+        public bool Current
+        {
+            get
+            {
+                return (_buffer[_currentBit >> 5] & (1 << (_currentBit & 0x1F))) != 0;
+            }
+        }
+
+        public bool MoveNext()
+        {
+            _currentBit++;
```

Comment:
OK, as I'm the king nitpicker today, calling this 2G times causes the _currentBit variable to overflow :D.

Reply:
The current uses of `BitEnumerator` don't hit this since the pattern is to use foreach, which terminates the first time `_currentBit` goes past `_limitBit`.

Reply:
Sure, I was just joking, I'm totally fine with that. I know it's typically perf-negative having to check the limit upfront in the various enumerator implementations and in the vast majority of cases the extra "diligence" in doing so doesn't bring any real benefit.